src/apply.c

```diff
@@ -376,3 +383,316 @@ int git_apply__patch(
 
 	return error;
 }
+
+static int apply_one(
+	git_repository *repo,
+	git_reader *preimage_reader,
+	git_index *preimage,
+	git_index *postimage,
+	git_diff *diff,
+	size_t i)
+{
+	git_patch *patch = NULL;
+	git_buf pre_contents = GIT_BUF_INIT, post_contents = GIT_BUF_INIT;
+	const git_diff_delta *delta;
+	char *filename = NULL;
+	unsigned int mode;
+	git_oid pre_id, post_id;
+	git_index_entry pre_entry, post_entry;
+	int error;
+
+	if ((error = git_patch_from_diff(&patch, diff, i)) < 0)
+		goto done;
+
+	delta = git_diff_get_delta(diff, i);
+
+	if (delta->status != GIT_DELTA_ADDED) {
+		error = git_reader_read(&pre_contents, &pre_id,
+			preimage_reader, delta->old_file.path);
+
+		/* ENOTFOUND means the preimage was not found; apply failed. */
+		if (error == GIT_ENOTFOUND)
+			error = GIT_EAPPLYFAIL;
+
+		/* When applying to BOTH, the index did not match the workdir. */
+		if (error == GIT_READER_MISMATCH)
+			error = apply_err("%s: does not match index", delta->old_file.path);
+
+		if (error < 0)
+			goto done;
+
+		if (preimage) {
+			memset(&pre_entry, 0, sizeof(git_index_entry));
+			pre_entry.path = delta->old_file.path;
+			pre_entry.mode = delta->old_file.mode;
+			git_oid_cpy(&pre_entry.id, &pre_id);
+
+			if ((error = git_index_add(preimage, &pre_entry)) < 0)
+				goto done;
+		}
```

Comment:
Just to make sure I understand: do you try to generate the preimage here, assuming that the passed `preimage` parameter is empty? I'm not a hundred percent sure I totally get that code

Reply:
I'll add a comment, but what we're doing here is populating a data structure that contains the actual on-disk contents as we read them out of the preimage - the actual on-disk contents of the file that we want to apply a patch to.

This is because the patch may be applied to a dirty working directory copy, so we need to know the expected baseline when we do the checkout, so that it can validate that the file hasn't changed.

eg, if file A is at abc1 in `HEAD`, and then modified in the working directory, we're still allowed to try to apply a patch.  So we need to get the hash of the contents of the file that we actually read and applied the patch to.  We'll provide that to checkout as the baseline, so it will verify that although the file is different than `HEAD`, it's what we actually expect it to be.  Without this, we'll have a checkout conflict.

Reply:
Added a comment to explain this.